### PR TITLE
Rewrite void_return rule using SwiftSyntax

### PR DESCRIPTION
**Pull request by marcelofabri**

### CODE REVIEW

**Review by jpsim — APPROVED** [review]
Lovely

File: Source/SwiftLintFramework/Rules/Style/VoidReturnRule.swift [hunk 1]
```diff
@@ -41,28 +40,83 @@ public struct VoidReturnRule: ConfigurationProviderRule, SubstitutionCorrectable
...
+            region.contains(node.positionAfterSkippingLeadingTrivia, locationConverter: locationConverter)
+        }
+
+        guard !isInDisabledRegion else {
+            return super.visit(node)
+        }
+
+        correctionPositions.append(tuple.positionAfterSkippingLeadingTrivia)
+
+        var returnType = SyntaxFactory.makeTypeIdentifier("Void")
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/Style/VoidReturnRule.swift:116]: TIL about `SyntaxFactory`

**Comment by jpsim** [comment]
@marcelofabri can we merge this?

**Comment by marcelofabri** [comment]
> @marcelofabri can we merge this?

I need to think about what to do with these violations. My current thinking is that it's fine to have them fixed here and update `redundant_void_return` to catch them, wdyt?

**Comment by jpsim** [comment]
Looking at the OSSCheck report, it looks like all the violations are safe because they're in closures with enough parent type context to infer the void return value. That's not the case with the current implementation of `redundant_void_return`.

**Comment by ZevEisenberg** [comment]
+1 I can't speak to the rewrite, but this fixes a case that currently incorrectly triggers on main, and which might be good to add to the test cases:

```swift
func foo() -> () async -> Int
```

Failure on main:

    testWithDefaultConfiguration(): failed - nonTriggeringExample violated: 
    ```
    /* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
    func foo() -> () async -> Int
                  ^ warning: Void Return Violation: Prefer `-> Void` over `-> ()`. (void_return)
    ```

**Comment by ZevEisenberg** [comment]
bump - I just rediscovered this thread because I dusted off an old project and found that this was still the case

**Comment by lo1tuma** [comment]
Is there anything missing or something I could help with to get this PR merged?

**Comment by SimplyDanny** [comment]
Rule has been rewritten in #5351.

**Comment by ZevEisenberg** [comment]
I'm not at a computer right now so I can't test it, but do you know if it now correctly handles the case I mentioned [above](https://github.com/realm/SwiftLint/pull/4158#issuecomment-1250129148)?

**Comment by SimplyDanny** [comment]
> I'm not at a computer right now so I can't test it, but do you know if it now correctly handles the case I mentioned [above](https://github.com/realm/SwiftLint/pull/4158#issuecomment-1250129148)?

Yes, the example is handled correctly now. See #5381.

**Pull request closed without merging**